### PR TITLE
Detect process.exit() called from tests

### LINKED ISSUE #861: Override `process.exit` to give some context when it's called unexpectedly in a test.
[Issuehunt badges](https://issuehunt.io/r/avajs/ava/issues/861)

Just had an issue on Gitter where a users code was causing `process.exit(1)` to be called. He got the "test.js exited with a non-zero exit code message", but there was no helpful information.

I say we try and capture that somehow:

``` js
var oldExit = process.exit;
process.exit = function (code) {
  if (code > 0) {
    var stack = new Error().stack;
    // log it? send it to parent?
    // ...
  }
  oldExit.apply(process, arguments);
};
```

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/avajs/ava/issues/861).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/ava/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by gibson042**
Fixes #861

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#861: Override `process.exit` to give some context when it's called unexpectedly in a test.](https://oss.issuehunt.io/repos/26820798/issues/861)
---
</details>

### CODE REVIEW

File: lib/worker/base.js [hunk 1]
```diff
@@ -19,6 +19,35 @@ import {flags, refs, sharedWorkerTeardowns} from './state.cjs';
...
+	if (!forceSync) {
+		await channel.flush();
+	}
+
+	apply(realExit, process, [code]);
+}
+
+process.exit = new Proxy(realExit, {
+	apply(fn, receiver, args) {
+		channel.send({type: 'internal-error', err: serializeError('Forced exit error', false, new Error('Unexpected process.exit()'), runner.file)});
```
Comment (novemberborn) [inline lib/worker/base.js:40]: Could leave `runner` undefined and use `runner?.file`, that should work in the supported Node.js versions I believe.

Comment (novemberborn) [inline lib/worker/base.js:40]: I think `internal` should refer to AVA doing something wrong, but this is triggered by user code. What do you think?

**Review by novemberborn — COMMENTED** [review]
Amazing!

File: lib/worker/base.js (hunk 1 above)
Comment (gibson042) [reply lib/worker/base.js]: Done.

Comment (gibson042) [reply lib/worker/base.js]: I think that's valid, although there is also an argument to be made that early termination before all tests run is an internal error. But what should this be instead?

Comment (novemberborn) [reply lib/worker/base.js]: `process-exit` maybe? Would also allow us to print a specific message, like we do for `SIGINT`:

https://github.com/avajs/ava/blob/392ab65cd4d203643e5bb29aa7f152056fb6040b/lib/reporters/default.js#L195-L200

Comment (gibson042) [reply lib/worker/base.js]: Done.

**Comment by novemberborn** [comment]
@gibson042 I should have some time this week to help get this over the line.

**Comment by novemberborn** [comment]
@gibson042 thanks for [`96ff6ef` (#3080)](https://github.com/avajs/ava/pull/3080/commits/96ff6ef0b80cba3d42d2237cb7e3d410ce9b8776)! Let's see if it improves things.

I've pushed several updates, please let me know what you think.

**Comment by gibson042** [comment]
@novemberborn The updates look good to me. My only further change was to rename the generic `proxyApply` to `handleProcessExit`.

**Pull request merged**